contracts/src/BatchManagerOperations.sol

```diff
@@ -0,0 +1,610 @@
+            _newBatchManager
+        );
+
+        vars.troveManager.onSetInterestBatchManager(
+            ITroveManager.OnSetInterestBatchManagerParams({
+                troveId: _troveId,
+                troveColl: vars.trove.entireColl,
+                troveDebt: vars.trove.entireDebt,
+                troveChange: newBatchTroveChange,
+                newBatchAddress: _newBatchManager,
+                newBatchColl: vars.newBatch.entireCollWithoutRedistribution,
+                newBatchDebt: vars.newBatch.entireDebtWithoutRedistribution
+            })
+        );
+
+        vars.sortedTroves.remove(_troveId);
+        vars.sortedTroves.insertIntoBatch(
+            _troveId,
+            BatchId.wrap(_newBatchManager),
+            vars.newBatch.annualInterestRate,
+            _upperHint,
+            _lowerHint
+        );
+    }
+
+    function kickFromBatch(
+        uint256 _troveId,
+        uint256 _upperHint,
+        uint256 _lowerHint
+    ) external {
+        _removeFromBatchInternal(
+            _troveId,
+            0, // ignored when kicking
+            _upperHint,
+            _lowerHint,
+            0, // will use the batch's existing interest rate, so no fee
+            true
+        );
+    }
+
+    function removeFromBatch(
+        uint256 _troveId,
+        uint256 _newAnnualInterestRate,
+        uint256 _upperHint,
+        uint256 _lowerHint,
+        uint256 _maxUpfrontFee
+    ) external {
+        _removeFromBatchInternal(
+            _troveId,
+            _newAnnualInterestRate,
+            _upperHint,
+            _lowerHint,
+            _maxUpfrontFee,
+            false
+        );
+    }
+
+    function _removeFromBatchInternal(
+        uint256 _troveId,
+        uint256 _newAnnualInterestRate,
+        uint256 _upperHint,
+        uint256 _lowerHint,
+        uint256 _maxUpfrontFee,
+        bool _kick
+    ) internal {
+        LocalVariables_removeFromBatch memory vars;
+        vars.troveManager = troveManager;
+        vars.sortedTroves = sortedTroves;
+
+        if (_kick) {
+            _requireTroveIsOpen(vars.troveManager, _troveId);
+        } else {
+            _requireTroveIsActive(vars.troveManager, _troveId);
+            _requireCallerIsBorrower(_troveId);
+            _requireValidAnnualInterestRate(_newAnnualInterestRate);
+        }
+
+        vars.batchManager = _requireIsInBatch(_troveId);
+        vars.trove = vars.troveManager.getLatestTroveData(_troveId);
+        vars.batch = vars.troveManager.getLatestBatchData(vars.batchManager);
+
+        if (_kick) {
+            if (
+                vars.batch.totalDebtShares * MAX_BATCH_SHARES_RATIO >=
+                vars.batch.entireDebtWithoutRedistribution
+            ) {
+                revert BatchSharesRatioTooLow();
+            }
+            _newAnnualInterestRate = vars.batch.annualInterestRate;
+        }
+
+        IBorrowerOperations(address(this)).removeTroveFromBatch(_troveId);
```

Comment:
nit: Was looking through the code, and we could get rid of this callback to the borrower operation and delete the storage mapping after the delegatecall.